transport/tcp.go

```diff
@@ -0,0 +1,102 @@
+package transport
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/hashicorp/memberlist"
+)
+
+// DialAddressTimeout is used by memberlist to establish a TCP connection to a particular node
+func (t *Transport) DialAddressTimeout(addr memberlist.Address, timeout time.Duration) (net.Conn, error) {
+	dialer := net.Dialer{
+		Timeout: timeout,
+	}
+	return dialer.Dial("tcp", addr.Addr)
+}
+
+// DialTimeout is used by memberlist to connect to a particular TCP Address
+func (t *Transport) DialTimeout(addr string, timeout time.Duration) (net.Conn, error) {
+	peer := memberlist.Address{
+		Addr: addr,
+	}
+	return t.DialAddressTimeout(peer, timeout)
+}
+
+// StreamCh is used by memberlist to receive incoming TCP connections
+func (t *Transport) StreamCh() <-chan net.Conn {
+	return t.streamCh
+}
+
+// tcpLoop is the main routine of the TCP listening workers
+func (t *Transport) tcpLoop(ctx context.Context, ln *net.TCPListener) {
+	// we explicitly close the listener because we could be interrupted
+	// by the cancellation of the parent Context instead of Shutdown()
+	defer ln.Close()
+
+	const baseDelay = 5 * time.Millisecond
+	const maxDelay = 1 * time.Second
+
+	var errorDelay time.Duration
+
+	for {
+		conn, err := ln.Accept()
+
+		if err == nil {
+			// no error, incoming connection
+			errorDelay = 0
+
+			t.debug().
+				WithField(ListenerAddrLabel, ln.Addr()).
+				WithField(RemoteAddrLabel, conn.RemoteAddr()).
+				Print("Connected")
+
+			select {
+			case t.streamCh <- conn:
+				// continue
+			case <-ctx.Done():
+				// sorry pal, we are cancelled
+				defer conn.Close()
+
+				t.error(ctx.Err()).
+					WithField(ListenerAddrLabel, ln.Addr()).
+					WithField(RemoteAddrLabel, conn.RemoteAddr()).
+					Print("Connection Terminated")
+
+				break
```

Comment:
does this ```break``` do really what you want? Didn't you mean:
```go
mainfor:for {
....
break mainfor
}
```

Reply:
good catch, but what about using `return` instead? @karasz